[AUDIO] [9.12] cirrus_sony: Enable backend mixer (for calibration) in a device agnostic manner

MarijnS95 opened the pull request:
Supersedes #19

Instead of hardcoding the mixer for `deep-buffer-playback` we might as well... select that path directly (second commit) or use the "usecase API" for `audio_extn` clients just like `spkr_protection.c` does (third commit).

CC @kholk

MarijnS95 committed:
cirrus_sony: Do not unpack unused function pointers from init config

These functions are not used; it's rather wasteful readability-wise to
retrieve them.

(The others are not used yet... but will be in future commits)

Signed-off-by: Marijn Suijten <marijns95@gmail.com>

MarijnS95 committed:
cirrus_sony: Enable deep buffer playback mixer through paths

Edo has the cirrus AMPs on a different backend (SEN_MI2S_RX to be exact,
instead of PRI on seine). The mixer listed here is to enable playback
through the deep-buffer usecase, so we can simply enable that path to
get the right backend enabled regardless of the device based on its
mixer_paths.

At the same time "clarify" hardcoded PCM id `0` by reading it from the
configuration (potentially overridden by audio_platform_info.xml)
instead.

Signed-off-by: Marijn Suijten <marijns95@gmail.com>

MarijnS95 committed:
cirrus_sony: Select usecase and enable backend through extn API

Enable silent audio playback (for calibration) through more verbose but
more specific APIs instead of manually poking around with the paths.

Signed-off-by: Marijn Suijten <marijns95@gmail.com>

MarijnS95 committed:
cirrus_sony: Wait until `primary_output` becomes available

This pointer strangely has a value immediately on Edo, but not on Seine.

MarijnS95 commented on hal/audio_extn/cirrus_sony.c:
```diff
@@ -600,17 +594,48 @@ static int cirrus_play_silence(int seconds) {
+    uc_info_rx->in_snd_device = SND_DEVICE_NONE;
+    uc_info_rx->stream.out = adev->primary_output;
+    list_init(&uc_info_rx->device_list);
+    // if (fp_audio_extn_is_vbat_enabled())
+    //     uc_info_rx->out_snd_device = SND_DEVICE_OUT_SPEAKER_PROTECTED_VBAT;
+    // else
+        // uc_info_rx->out_snd_device = SND_DEVICE_OUT_SPEAKER_PROTECTED;
+    uc_info_rx->out_snd_device = SND_DEVICE_OUT_SPEAKER;
```
What even does vbat mean... And do we want the protected speaker here? Neither works...

kholk replied:
VBAT means exactly what you think it means. VBAT. Battery voltage.

MarijnS95 replied:
What do speakers have to do with battery voltage? Compensation for decreasing voltage and hence volume?

MarijnS95 replied:
Anyway, which of the 3 should we use @kholk?

kholk reviewed (approved):
I had left them because I had a plan, but that's probably not necessary.

Since you said about "the others", then you know what my plan was.... :P

kholk commented on hal/audio_extn/cirrus_sony.c:
```diff
@@ -597,15 +594,13 @@ static int cirrus_play_silence(int seconds) {
     unsigned int buffer_size = 0, frames_bytes = 0;
+    int pcm_dev_rx_id = fp_platform_get_pcm_device_id(USECASE_AUDIO_PLAYBACK_DEEP_BUFFER, PCM_PLAYBACK);
 
+    audio_route_apply_and_update_path(adev->audio_route, "deep-buffer-playback");
```
That's a bit more neat than what I was imagining to do.
In my head, I was ready to grab it from some imaginary property or something like that.

Kudos.

kholk commented on hal/audio_extn/cirrus_sony.c:
```diff
@@ -592,13 +594,46 @@ static int cirrus_play_silence(int seconds) {
     struct mixer_ctl *ctl_config = NULL;
     struct pcm_config rx_tmp = { 0 };
+    struct audio_usecase *uc_info_rx;
+    #define CIRRUS_PLAYBACK_MIXER "SEN_MI2S_RX Audio Mixer MultiMedia1"
```
Sorry, your development notes are not allowed... :P :P :P :P :P

MarijnS95 replied:
Lmao that's remaining from a readback from this thing, to show that is actually getting activated. Gross!

kholk replied:
Yeah, I also do that rather frequently. I know :D

kholk commented on hal/audio_extn/cirrus_sony.c:
```diff
@@ -611,6 +611,13 @@ static int cirrus_play_silence(int seconds) {
+    while (!adev->primary_output) {
+        ALOGE("Still no primary_output!");
+        // TODO: Perhaps wait on a condvar like spkr_prot?
+        usleep(1000);
```
Are you sure that this really works? Is this stuff being executed in a pthread while the HAL continues the other part of the init?!??!?

Because in my head again, the solution for this was to spawn a silence-calibration pthread (but that was WAAAAAAAAAY MORE complicated than what you just did, you know, sync everything, mutexes blah, make sure that nothing plays before this function succeeds blahblah, big mess)

MarijnS95 replied:
Why don't we take a look in `spkr_protection`?

https://github.com/sonyxperiadev/vendor-qcom-opensource-audio-hal-primary-hal/blob/13b245e5edfadcb235e2ac32c61b0eee6328c74d/hal/audio_extn/spkr_protection.c#L1275-L1278

https://github.com/sonyxperiadev/vendor-qcom-opensource-audio-hal-primary-hal/blob/13b245e5edfadcb235e2ac32c61b0eee6328c74d/hal/audio_extn/spkr_protection.c#L1123-L1131

Hmm, so we have a condvar...

That is only ever signaled here:

https://github.com/sonyxperiadev/vendor-qcom-opensource-audio-hal-primary-hal/blob/13b245e5edfadcb235e2ac32c61b0eee6328c74d/hal/audio_extn/spkr_protection.c#L1930-L1935

That's only ever called from deinit... and...

https://github.com/sonyxperiadev/vendor-qcom-opensource-audio-hal-primary-hal/blob/13b245e5edfadcb235e2ac32c61b0eee6328c74d/hal/audio_extn/spkr_protection.c#L2014

Which is called from

https://github.com/sonyxperiadev/vendor-qcom-opensource-audio-hal-primary-hal/blob/13b245e5edfadcb235e2ac32c61b0eee6328c74d/hal/audio_extn/audio_extn.c#L6570

MarijnS95 replied:
By the way, notice that `cirrus_do_calibration` already runs asynchronously in a thread.

kholk commented:
Basically, as a global comment to the PR, the only thing to fix is a development leftover (one line).

Apart from this, Reviewed-By: .... :)))

MarijnS95 committed:
cirrus_sony: Use tabs instead of spaces

MarijnS95 committed:
cirrus_sony: Do not sum error codes with positive string-length

cirrus_format_mixer_name returns the number of formatted characters
(result of snprintf). Returning this positive number is not always a
good idea, like when check_error_state returns 22 as the length of a
mixer control, which is caught by the `==0` check resulting in a full
reset.
Secondly, a postive length of 22 together with an actual, expected
failure of -EINVAL which conveniently is -22 results in 0; no error at
all!!

Signed-off-by: Marijn Suijten <marijns95@gmail.com>

kholk committed:
cirrus_sony: Redo PCM handling for calibration, add locking

There is no need (and it's wrong) to call pcm_start before
writing the data, as we can simply retrieve, prepare and
write the silence data to the PCM instead.

Also, while at it, add some locking around to make sure that
we don't race while setting module states, especially in
corner cases where we need to calibrate the speakers before
starting processing (very, very unlikely case that should
anyway never happen).

Co-authored-by: Marijn Suijten <marijn.suijten@somainline.org>

kholk committed:
cirrus_sony: Fixes for new firmware mixer names and order

The new firmware features more creativity, which means that the
mixer names have changed. Also, for some reason, when we've got
a dual amp setup, the DSP refuses to load the R-SPK firmware by
caprice if we load the L-SPK one before that, even though they
definitely have no real correlation between each other (or at
least not firmware-wise).

Even though there are *way* better ways to do this, time does
not allow me to, so this is a very fast workaround to get the
audio calibrations kicking.

kholk committed:
cirrus_sony: Consolidate calibration process

On devices with multiple CS35L41 amps the calibration process is a bit
slower for.. reasons.
First of all, the DSP likes to just take time to get back alive after
the calibration run succeeding (reasons unknown: firmware is closed)
and there is no way to know when it precisely comes back to life.

Then, this task being now very much time consuming brings more issues
to the table, like the audio HAL trying to playback sounds from the
OS before the calibration run is done: this may actually interact with
the calibration silence and/or confuse the DSP and make it crash, which
is why we now need to carefully set and check the module state, meaning
that we are obliged to check and forbid audio playback if this software
is waiting for DSP calibration.

The result of this is that *at least* Sony Edo, featuring this kind of
dual-amp cofiguration, can now calibrate and switch to protection FW
without requiring a system reboot, even though it's going to be a bit
slow: this was ignored for the sake of safety and audio quality through
the internal speakers and it is very important to recalibrate this
when we have the opportunity to, as the TA-stored calibration cannot be
accounting for hardware aging while, with this, WE CAN.

kholk committed:
cirrus_sony: Loop around DSP state after calibration

Instead of waiting for a arbitrary amount of time in any case,
a better user experience may be (in some cases) achieved by
looping through the DSP state: if we succeed to change the DSP
firmware mixer (to load the new FW) before the max wait, that's
going to be... well... less time to lose.

This actually brings benefits depending on the specific device
and on the specific boot: for example, my PDX203 sometimes
takes as much as 5 seconds, sometimes just two, and this probably
depends on the boot order since it's entirely async, which gets
different depending on the device state before actually booting,
as the hardware will be in different states and some daemons
init time will be skewed;

For example:
- Bootloader version X will leave some HW in a different state
  than version Y; or
- The device is booting from a fastboot continue command; or
- The device is cold booting.

MarijnS95 committed:
cirrus_sony: Ignore start_processing on the calibration thread

We use fp_enable_snd_device (and friends) to set up the right
routes/paths and backends before trying to use the speakers, with the
annoying caveat that start_processing is called. Usually after a reboot
this only spews `Forbidden. Calibration is in progess...` once, but if
AOSP already tries playing audio it'll get stuck in a loop with lots of
pops and clicks.

Signed-off-by: Marijn Suijten <marijns95@gmail.com>

kholk commented:
Looks perfect. Remove the WIP status and MERGE IT ASAP!!!

MarijnS95 commented:
I agree. The CAF HAL is horrible and we still have some weird issue when audio already is already playing (ie. a running media player) but calibration still has to occur; that's a unique situation that can really only happen while debugging.

Let's test this on Seine ITMT... The previous PR was working but the new patches are breaking "somewhere".

MarijnS95 commented on hal/audio_extn/cirrus_sony.c:
```diff
@@ -600,35 +600,64 @@ static int cirrus_play_silence(int seconds) {
+    uc_info_rx->in_snd_device = SND_DEVICE_NONE;
+    uc_info_rx->stream.out = adev->primary_output;
+    list_init(&uc_info_rx->device_list);
+    uc_info_rx->out_snd_device = SND_DEVICE_OUT_SPEAKER_PROTECTED;
```
@kholk I've changed this to protected again. Normal speaker doesn't work on Seine (`pcm_write` returns `-1`), and we've seen before that the HAL maps speaker to speaker_protected in _most_ scenarios _anyway_ because we're a "spkr prot".

It's probably not doing that translation everywhere, leading to a mismatch and breakage. But I don't think either of us cares enough to debug. `SND_DEVICE_OUT_SPEAKER_PROTECTED` also works on edo.
